promql/functions.go

```diff
@@ -130,6 +134,209 @@ func extrapolatedRate(vals []parser.Value, args parser.Expressions, enh *EvalNod
+				counterCorrection += lastValue //** Accumulate the sum of backward steps
+			}
+			lastValue = sample.V
+		}
+	}
+	resultValue := points[len(points)-1].V - points[firstPoint].V + counterCorrection //** last.V - first.V + backward correction
+
+	// Duration between last sample and boundary of range.
+	durationToEnd := float64(rangeEnd - points[len(points)-1].T)
+
+	// If the points cover the whole range (i.e. they start just before the
+	// range start and end just before the range end) adjust the value from
+	// the sampled range to the requested range.
+	if points[firstPoint].T <= rangeStart && durationToEnd < averageInterval {
+		adjustToRange := float64(durationMilliseconds(ms.Range))
+		resultValue *= (adjustToRange / sampledRange)
+	}
+
+	if isRate {
+		resultValue /= ms.Range.Seconds()
+	}
+
+	return append(enh.Out, Sample{
+		Point: Point{V: resultValue},
+	})
+}
+
+// Infers the scrape interval in msec by taking the median interval among the first 25.
+// By using the median, we avoid the highs (late or missed scrape) and the lows
+// (where a late scrape caused the next to be too close).
+func inferScrapeInterval(points []Point) int64 {
+	if len(points) == 0 {
+		panic(errors.Errorf("no points to work on"))
+	}
+
+	// Compute up to 25 time intervals and store them in intervalArray.
+	var intervalArray [25]int64
+	var i int
+	for i = 0; i < (len(points)-1) && i < len(intervalArray); i++ {
+		intervalArray[i] = points[i+1].T - points[i].T
+	}
+
+	// Slice and sort the intervals (may be fewer than 25, if points was shorter than 26).
+	intervals := intervalArray[:i]
+	sort.Slice(intervals, func(i0, i1 int) bool { return intervals[i0] < intervals[i1] })
+
+	// Return the median interval.
+	return intervals[len(intervals)/2]
+}
+
+const elideSamplesAfter int = 10
+
+// extendedRate0 is a utility function for xrate0/xincrease0.
+// It calculates the rate (allowing for counter resets and including the startup value from 0),
+// taking into account the last sample before the range start, and returns
+// the result as either per-second (if isRate is true) or overall.
+func extendedRate0(vals []parser.Value, args parser.Expressions, enh *EvalNodeHelper, isRate bool) Vector {
+	ms := args[0].(*parser.MatrixSelector)
+	vs := ms.VectorSelector.(*parser.VectorSelector)
+
+	rangeStartMsec := enh.Ts - durationMilliseconds(ms.Range+vs.Offset)
+	rangeEndMsec := enh.Ts - durationMilliseconds(vs.Offset)
+
+	samples := vals[0].(Matrix)[0]
+	points := samples.Points
+
+	// Cannot compute rate with 0 or 1 data points.
+	if len(points) < 2 {
+		return enh.Out
+	}
+
+	scrapeIntervalMsec := inferScrapeInterval(points) //** milliseconds
+
+	firstPoint := 0
+	var lastValue float64
+
+	// The 0 Fix: Check for fresh pod start case where we have no point before the range,
+	// meaning the caller failed to find any before the range in the lookback interval.
+	// Treat this as if we had a 0 to start.
+	firstPointToRangeStartMsec := rangeStartMsec - points[0].T
```

Comment:
Note the 2 special cases handled below. These took a while to get right.
1. No point before the range. So assume a fresh start and inject 0 as the `lastValue`.
2. There is a point before the range, but it's _too_ far before. So skip it.
otherwise we have a normal case, with a point (just) before the range for context.